server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+        val res = for {
+          instance <- Task.delay(Mac.getInstance(CSRFMiddleware.SigningAlgo))
+          _ <- Task.delay(instance.init(config.key))
+          out <- Task.delay(
+            instance.doFinal(
+              (raw + "-" + nonce).getBytes(StandardCharsets.UTF_8)))
+          r <- Task.point(
+            MessageDigest
+              .isEqual(out, Base64.getDecoder.decode(signed))
+              .option(raw))
+        } yield r
+        OptionT(res)
+      case _ =>
+        OptionT.none
+    }
+
+  /** Constructs a middleware that will check for the csrf token
+    * presence on both the proper cookie, and header values.
+    *
+    * If it is a valid token, it will then embed a new one,
+    * to effectively randomize the complete token while
+    * avoiding the generation of a new secure random Id, to guard
+    * against [BREACH](http://breachattack.com/)
+    *
+    */
+  def validate: Middleware[Request, MaybeResponse, Request, MaybeResponse] = {
+    service =>
+      Kleisli[Task, Request, MaybeResponse] { r =>
+        (for {
+          c1 <- CSRFMiddleware.cookieFromHeaders(r, config.cookieName)
+          c2 <- OptionT(
+            Task.point(r.headers.get(CaseInsensitiveString(config.headerName))))
+          raw1 <- extractRaw(c1.content)
+          raw2 <- extractRaw(c2.value)
+          response <- if (CSRFMiddleware.isEqual(raw1, raw2)) {
+            OptionT[Task, MaybeResponse](service(r).map(Some(_)))
+          } else {
+            OptionT.none[Task, MaybeResponse]
+          }
+          newToken <- OptionT[Task, String](signToken(raw1).map(Some(_)))
+        } yield
+          response.cata(
+            _.addCookie(Cookie(config.cookieName, newToken)),
+            Response(Status.NotFound)))
+          .getOrElse(Response(Status.Unauthorized))
+      }
+  }
+
+  /** Embed a token into a response **/
+  def embedNew(res: MaybeResponse): Task[Response] =
+    generateToken.map(
+      content =>
+        res.cata(_.addCookie(Cookie(config.cookieName, content)),
+          Response(Status.NotFound)))
+
+  /** Middleware to embed a csrf token into routes that do not have one. **/
+  def withNewToken: Middleware[Request, MaybeResponse, Request, MaybeResponse] =
+    _.andThen(Kleisli(embedNew))
+
+}
+
+object CSRFMiddleware {
+
+  /** An instance of SecureRandom to generate
+    * tokens, properly seeded:
+    * https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/
+    */
+  private lazy val CachedRandom = {
+    val r = new SecureRandom()
+    r.nextBytes(new Array[Byte](20))
+    r
+  }
+
+  val SigningAlgo = "HmacSHA1"
+  val SHA1ByteLen = 20
+  val CSRFTokenLength = 32
+
+  /** Hex encoding digits. Adapted from apache commons Hex.encodeHex **/
+  private val Digits: Array[Char] = Array('0', '1', '2', '3', '4', '5', '6',
+    '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  protected def find[A](nel: NonEmptyList[A], p: A => Boolean): Option[A] =
```

Comment:
We use our own `NonEmptyList` in 0.16, because Scalaz broke theirs across versions.  I think we could put this straight on there.

We have syntax for NonEmptyList in later versions to add the operations that are missing from cats'.  When it comes time to merge up there, I think this belongs there.  I also think we should offer these missing methods to cats.

Reply:
this `find` is already in cats actually. I can add it to the http4s nel.